README.md
leapseconds-announced
=====================

This libraray provides a static [`Data.Time.Clock.TAI.LeapSecondTable`][1]
"containing" the leap seconds announced at library release time.
A new version of the library is released every time the International
Earth Rotation and Reference Systems Service (IERS) announces a new
leap second at http://hpiers.obspm.fr/eoppc/bul/bulc/bulletinc.dat.

This module is intended to provide a quick-and-dirty leap second solution
for one-off analyses concerned only with the past and present (i.e. up
until the next as of yet unannounced leap second), or for applications
which can afford to be recompiled against an updated library as often
as every six months.

[1]: https://hackage.haskell.org/package/time/docs/Data-Time-Clock-TAI.html


Usage
-----

Import `Data.Time.Clock.AnnouncedLeapSeconds` to bring the leap
second table `lst` into scope. Here is a usage example:

```haskell
import Data.Time
import Data.Time.Clock.TAI
import Data.Time.Clock.AnnouncedLeapSeconds

-- | Convert from UTC to TAI.
utcToTAITime' = utcToTAITime lst
-- | Convert from TAI to UTC.
taiToUTCTime' = taiToUTCTime lst

-- | Add a length of time to a UTC time, respecting leap seconds
--   (as opposed to Data.Time.Clock.addUTCTime).
addUTCTime' :: DiffTime -> UTCTime -> UTCTime
addUTCTime' dt = taiToUTCTime' . addAbsoluteTime dt . utcToTAITime'
```


changelog.md
2017.1 (2016-12)
====================

*  Corrected version numbering to reflect API breaking (2017.0.0.1 deprecated).

2017.0.0.1 (2016-12)
====================

*  Ported leapSecondMap of time-1.7.

2017 (2016-06)
==============

*  Added 2017 leap second (2016-12-31T23:59:60Z).

2015.0.0.1 (2015-05)
====================

*  Fix documentation to reflect addition of 2015 leap second.
*  Hide hypothetical leap seconds (prior to 1972-01-01).
*  Add test suite to .cabal.

2015 (2015-01)
==============

*  Added 2015 leap second (2015-06-30T23:59:60Z).

2012 (2012-01)
==============

*  Added 2012 leap second (2012-06-30T23:59:60Z).

2009 (2009-01)
==============

Initial release including all leap seconds up to 2009-01-01.


tests/Test.hs
import Data.Time
import Data.Time.Clock.AnnouncedLeapSeconds
import Test.QuickCheck

onceCheck :: Bool -> IO ()
onceCheck = quickCheckWith (stdArgs {maxSuccess = 1})

-- A few trivial tests.
main :: IO ()
main = do
  onceCheck $ lst (fromGregorian 1111 12 31) ==  0  -- Before first leap second.
  onceCheck $ lst (fromGregorian 2008 12 31) == 33
  onceCheck $ lst (fromGregorian 2009 01 01) == 34
  onceCheck $ lst (fromGregorian 2009 12 31) == 34
  onceCheck $ lst (fromGregorian 2012 06 30) == 34  -- Prior to last leap second.
  onceCheck $ lst (fromGregorian 2012 07 01) == 35  -- Last leap second.
  onceCheck $ lst (fromGregorian 2012 12 31) == 35  -- Beyond last leap second.



MakeLeapSecondTable.hs
{- |
Use this applicaton to generate the 'Data.Time.Clock.AnnouncedLeapSeconds'
module. Compile and pipe an EOP file from Celestrak through the binary,
e.g.:

  curl http://www.celestrak.com/SpaceData/eop19620101.txt | ./MakeLeapSecondTable > Data/Time/Clock/AnnouncedLeapSeconds.hs

-}

import Astro.Celestrak
import Data.List (intercalate)
import Data.Time (Day)
import Data.Time.Format (formatTime)
import System.Locale (defaultTimeLocale)


-- | Converts an 'EOPList' into a minimal list of (day, leapsecond) pairs
-- in reverse chronological order.
eopToLS :: EOPList a -> [(Day, Integer)]
eopToLS = reverse . keepInitial . fmap (fmap deltaAT)

-- | Keeps the first pair with a given snd value while dropping the
-- following pairs with the same snd value.
keepInitial :: Eq a => [(b,a)] -> [(b,a)]
keepInitial (x:xs) = x : keepInitial (dropWhile (sndEq x) xs) where sndEq (_,x) (_,y) = x == y
keepInitial [] = []

{-
The above and this function are candidates for moving to "Astro.Celestrak":

-- | Converts an 'EOPList' to a light weight 'LeapSecondTable' (its internal
-- data is a short list as opposed to a huge array for the 'LeapSecondTable'
-- provided by "Astro.Celestrak".
eopToLST :: EOPList a -> LeapSecondTable
eopToLST eops d = snd $ headDef (undefined,0) $ dropWhile ((>d).fst) $ eopToLS eops
-}

-- | Convert a day/leapsecond pair into a compilable string.
lsToString :: (Day, Integer) -> String
lsToString (d,s) = formatTime defaultTimeLocale fmt d
  where fmt = "(fromGregorian %Y %m %d, " ++ show s ++ ")"

-- | Shows a list in compilable format using the passed function to display
-- the elements of the list.
showL :: (a -> String) -> [a] -> String
showL showf xs = intercalate "\n  : " (map showf xs) ++ "\n  : []"

-- | Compilable leapsecond module.
showModule :: EOPList a -> String
showModule eops = unlines
  [ "-- This file was automatically generated."
  , ""
  , "{- |"
  , "   Copyright  : Copyright (C) 2009-2015 Bjorn Buckwalter"
  , "   License    : BSD3"
  , ""
  , "   Maintainer : bjorn.buckwalter@gmail.com"
  , "   Stability  : stable"
  , "   Portability: full"
  , ""
  , "Provides a static 'Data.Time.Clock.TAI.LeapSecondTable' \\\"containing\\\""
  , "the leap seconds announced at library release time. This version"
  , "will become invalidated when/if the International Earth Rotation"
  , "and Reference Systems Service (IERS) announces a new leap second at"
  , "<http://hpiers.obspm.fr/eoppc/bul/bulc/bulletinc.dat>."
  , "At that time a new version of the library will be released, against"
  , "which any code wishing to remain up to date should be recompiled."
  , ""
  , "This module is intended to provide a quick-and-dirty leap second solution"
  , "for one-off analyses concerned only with the past and present (i.e. up"
  , "until the next as of yet unannounced leap second), or for applications"
  , "which can afford to be recompiled against an updated library as often"
  , "as every six months."
  , "-}"
  , ""
  , "module Data.Time.Clock.AnnouncedLeapSeconds (lst) where"
  , ""
  , "import Data.Maybe (listToMaybe)"
  , "import Data.Time (Day, fromGregorian)"
  , "import Data.Time.Clock.TAI (LeapSecondTable)"
  , ""
  , "-- | List of all leap seconds up to 2015-07-01. An"
  , "-- estimate of hypothetical leap seconds prior to 1972-01-01 is"
  , "-- included. These can be understood as leap seconds that may have"
  , "-- been introduced had UTC used the SI second since its inception in 1961."
  , "-- One should be extremely careful in using this information as it is"
  , "-- generally not appropriate. One specific case where it may be useful"
  , "-- is in reducing the error in computed time differences between UTC time"
  , "-- stamps in the 1961--1971 range from the order of 10 SI seconds to 1 SI"
  , "-- second."
  , "pseudoLeapSeconds :: [(Day, Int)]"
  , "pseudoLeapSeconds = " ++ showL lsToString ls
  , ""
  , "-- | List of all official leap seconds from 1972-01-01 to 2015-07-01."
  , "leapSeconds :: [(Day, Int)]"
  , "leapSeconds = takeWhile (> introduction) pseudoLeapSeconds ++ [introduction]"
  , "  where"
  , "    introduction = (fromGregorian 1972 01 01, 10)"
  , ""
  , "-- | 'Data.Time.Clock.TAI.LeapSecondTable' containing all leap seconds"
  , "-- from 1972-01-01 to " ++ (show.fst.head) ls ++ "."
  , "lst :: LeapSecondTable"
  , "lst d = fmap snd $ listToMaybe $ dropWhile ((>d).fst) leapSeconds"
  ] where ls = eopToLS eops

main = do
  interact (showModule . parseEOPData)


src/Data/Time/Clock/AnnouncedLeapSeconds.hs
-- This file was automatically generated.

{- |
   Copyright  : Copyright (C) 2009-2016 Bjorn Buckwalter
   License    : BSD3

   Maintainer : bjorn.buckwalter@gmail.com
   Stability  : stable
   Portability: full

Provides a static 'Data.Time.Clock.TAI.LeapSecondTable' \"containing\"
the leap seconds announced at library release time. This version
will become invalidated when/if the International Earth Rotation
and Reference Systems Service (IERS) announces a new leap second at
<http://hpiers.obspm.fr/eoppc/bul/bulc/bulletinc.dat>.
At that time a new version of the library will be released, against
which any code wishing to remain up to date should be recompiled.

This module is intended to provide a quick-and-dirty leap second solution
for one-off analyses concerned only with the past and present (i.e. up
until the next as of yet unannounced leap second), or for applications
which can afford to be recompiled against an updated library as often
as every six months.
-}

module Data.Time.Clock.AnnouncedLeapSeconds (lst) where

import Data.Time (Day, fromGregorian)
import Data.Time.Clock.TAI (LeapSecondTable)

-- | List of all leap seconds up to 2017-01-01. An
-- estimate of hypothetical leap seconds prior to 1972-01-01 is
-- included. These can be understood as leap seconds that may have
-- been introduced had UTC used the SI second since its inception in 1961.
-- One should be extremely careful in using this information as it is
-- generally not appropriate. One specific case where it may be useful
-- is in reducing the error in computed time differences between UTC time
-- stamps in the 1961--1971 range from the order of 10 SI seconds to 1 SI
-- second.
pseudoLeapSeconds :: [(Day, Integer)]
pseudoLeapSeconds = (fromGregorian 2017 01 01, 37)
  : (fromGregorian 2015 07 01, 36)
  : (fromGregorian 2012 07 01, 35)
  : (fromGregorian 2009 01 01, 34)
  : (fromGregorian 2006 01 01, 33)
  : (fromGregorian 1999 01 01, 32)
  : (fromGregorian 1997 07 01, 31)
  : (fromGregorian 1996 01 01, 30)
  : (fromGregorian 1994 07 01, 29)
  : (fromGregorian 1993 07 01, 28)
  : (fromGregorian 1992 07 01, 27)
  : (fromGregorian 1991 01 01, 26)
  : (fromGregorian 1990 01 01, 25)
  : (fromGregorian 1988 01 01, 24)
  : (fromGregorian 1985 07 01, 23)
  : (fromGregorian 1983 07 01, 22)
  : (fromGregorian 1982 07 01, 21)
  : (fromGregorian 1981 07 01, 20)
  : (fromGregorian 1980 01 01, 19)
  : (fromGregorian 1979 01 01, 18)
  : (fromGregorian 1978 01 01, 17)
  : (fromGregorian 1977 01 01, 16)
  : (fromGregorian 1976 01 01, 15)
  : (fromGregorian 1975 01 01, 14)
  : (fromGregorian 1974 01 01, 13)
  : (fromGregorian 1973 01 01, 12)
  : (fromGregorian 1972 07 01, 11)
  : (fromGregorian 1971 08 03, 10)
  : (fromGregorian 1970 07 13, 9)
  : (fromGregorian 1969 06 23, 8)
  : (fromGregorian 1968 06 02, 7)
  : (fromGregorian 1967 04 04, 6)
  : (fromGregorian 1966 03 15, 5)
  : (fromGregorian 1965 01 01, 4)
  : (fromGregorian 1963 08 07, 3)
  : (fromGregorian 1962 01 01, 2)
  : []

-- | List of all official leap seconds from 1972-01-01 to 2017-01-01.
leapSeconds :: [(Day, Integer)]
leapSeconds = takeWhile (> introduction) pseudoLeapSeconds ++ [introduction]
  where
    introduction = (fromGregorian 1972 01 01, 10)

-- | 'Data.Time.Clock.TAI.LeapSecondTable' containing all leap seconds
-- from 1972-01-01 to 2017-01-01.
lst :: LeapSecondTable
lst d = snd $ headDef (undefined,0) $ dropWhile ((>d).fst) leapSeconds
  where headDef def xs = if null xs then def else head xs  -- Inspired by Safe.
